### PR TITLE
fix: columns title render not as expected

### LINKED ISSUE #38371: Table的title属性使用函数和直接传入ReactNode行为不一致
### Reproduction link

[Edit on CodeSandbox](https://codesandbox.io/s/ji-ben-yong-fa-antd-4-24-0-forked-bgjhv3?file=/demo.tsx)

### Steps to reproduce

列配置的title如果是ReactNode,只取子一个子元素，如果是 () => ReactNode, 则会取整个

image

### What is expected?

都去整个ReactNode 作为title

### What is actually happening?

表现不一致

| Environment | Info |
| --- | --- |
| antd | undefined |
| React | 17.0.2 |
| System | macOS |
| Browser | chrome |

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [X] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 💡 Background and solution

- fix: #38371

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | columns title render not as expected |
| 🇨🇳 Chinese | 列表题渲染与预期不符 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

### CODE REVIEW

File: components/table/hooks/useSorter.tsx [hunk 1]
```diff
@@ -209,12 +209,12 @@ function injectSorter<RecordType>(
...
             cell['aria-sort'] = sorterOrder === 'ascend' ? 'ascending' : 'descending';
           } else {
-            cell['aria-label'] = `${renderColumnTitle(column.title, {})} sortable`;
+            cell['aria-label'] = `${renderColumnTitle(column.title, {}, true)} sortable`;
           }
           cell.className = classNames(cell.className, `${prefixCls}-column-has-sorters`);
           cell.tabIndex = 0;
           if (column.ellipsis) {
-            cell.title = (renderColumnTitle(column.title, {}) ?? '').toString();
+            cell.title = (renderColumnTitle(column.title, {}, true) ?? '').toString();
```
Comment (afc163) [inline components/table/hooks/useSorter.tsx:217]: 单开一个函数吧，这样调用语义清晰一点。

**Comment by afc163** [comment]
1. 现有的用例挂了
2. 需要再补一个用例 https://github.com/ant-design/ant-design/pull/38377

File: components/table/hooks/useSorter.tsx (hunk 1 above)
Comment (kiner-tang) [reply components/table/hooks/useSorter.tsx:217]: 好的

**Comment by afc163** [comment]
覆盖率掉了，这一行没跑到。

https://app.codecov.io/gh/ant-design/ant-design/pull/38378

图片

**Comment by kiner-tang** [comment]
> 覆盖率掉了，这一行没跑到。
> 
> https://app.codecov.io/gh/ant-design/ant-design/pull/38378
> 
> 图片

已经在优化测试用例了

File: components/table/util.ts [hunk 2]
```diff
@@ -33,19 +33,26 @@ function getElementFirstTextContent(node: React.ReactElement): string {
...
 
+/**
+ * Render title by first content
+ *
+ * @param title
+ * @returns
+ */
+export function renderColumnTitleWithFirstContent<RecordType>(title: ColumnTitle<RecordType>) {
+  if (React.isValidElement(title)) {
+    return getElementFirstTextContent(title);
```
Comment (zombieJ) [inline components/table/util.ts:44]: `getElementFirstTextContent` for check children is not a goodway which is easy to break on `HOC`:
```tsx
const MyTitle = () => 'title';

columns=[{
  title: <MyTitle />,
}]
```

**Review by afc163 — APPROVED** [review]

File: components/table/__tests__/__snapshots__/Table.test.tsx.snap [hunk 3]
```diff
@@ -204,3 +204,303 @@ exports[`Table rtl render component should be rendered correctly in RTL directio
...
+                              </svg>
+                            </span>
+                          </span>
+                        </span>
+                      </div>
+                    </th>
+                    <th
+                      aria-label="function title() {
+        return /*#__PURE__*/_react[\\"default\\"].createElement(\\"div\\", null, /*#__PURE__*/_react[\\"default\\"].createElement(\\"span\\", null, \\"name\\"), /*#__PURE__*/_react[\\"default\\"].createElement(\\"span\\", null, \\"Red\\"));
+      } sortable"
```
Comment (afc163) [inline components/table/__tests__/__snapshots__/Table.test.tsx.snap:370]: 这里应该是有问题的。

Comment (kiner-tang) [reply components/table/__tests__/__snapshots__/Table.test.tsx.snap:370]: 是的，需要处理一下函数类型的情况

Comment (afc163) [reply components/table/__tests__/__snapshots__/Table.test.tsx.snap:370]: 我们先回滚应急，原问题重新 PR 解决。

Comment (Wxh16144) [reply components/table/__tests__/__snapshots__/Table.test.tsx.snap:370]: 假设我们开发者这样去定义:

```jsx
const MyTitle = () => (
  <>
    <span style={{ color: 'red' }}>*</span>
    <span>name</span>
  </>
);
```

那么 `aria-label` 可能还是不符合预期

Comment (afc163) [reply components/table/__tests__/__snapshots__/Table.test.tsx.snap:370]: aria-label 应该交给用户去自定义。我们默认生成的 aria-label 可以是空的，不要是错的就行。

**Comment by afc163** [comment]
https://github.com/ant-design/ant-design/pull/38383

可以重新提个 PR 解决之前的问题。

**Comment by kiner-tang** [comment]
> #38383
> 
> 可以重新提个 PR 解决之前的问题。

- https://github.com/ant-design/ant-design/pull/38389

**Pull request closed without merging**